From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Sat, 29 Jan 2022 20:36:34 +0100
Subject: multitasking-panel: Sync workspace and monitor isolation dock
 settings

When using ubuntu-dock the multitasking panel setting doesn't fully
apply, because even though the user selects to isolate workspaces, the
dock may still continue to show icons from all the workspaces.

As per this, ensure that the settings are applied also to the dock, plus
include dock-specifc settings to control multi-monitor behavior.

Forwarded: not-need
---
 panels/multitasking/cc-multitasking-panel.c  | 56 ++++++++++++++++++++++++++++
 panels/multitasking/cc-multitasking-panel.ui | 35 +++++++++++++++++
 2 files changed, 91 insertions(+)

--- a/panels/multitasking/cc-multitasking-panel.c
+++ b/panels/multitasking/cc-multitasking-panel.c
@@ -31,6 +31,7 @@ struct _CcMultitaskingPanel
   GSettings       *interface_settings;
   GSettings       *mutter_settings;
   GSettings       *shell_settings;
+  GSettings       *dock_settings;
   GSettings       *wm_settings;
 
   CcIllustratedRow *active_screen_edges_row;
@@ -39,6 +40,8 @@ struct _CcMultitaskingPanel
   GtkCheckButton  *current_workspace_radio;
   GtkCheckButton  *dynamic_workspaces_radio;
   GtkCheckButton  *fixed_workspaces_radio;
+  GtkCheckButton  *dock_monitors_isolation_radio;
+  GtkCheckButton  *dock_each_monitor_radio;
   CcIllustratedRow *hot_corner_row;
   GtkSwitch       *hot_corner_switch;
   AdwSpinRow      *number_of_workspaces_spin_row;
@@ -50,6 +53,24 @@ struct _CcMultitaskingPanel
 CC_PANEL_REGISTER (CcMultitaskingPanel, cc_multitasking_panel)
 
 static void
+keep_dock_settings_in_sync (CcMultitaskingPanel *self)
+{
+  gboolean switcher_isolate_workspaces;
+  gboolean dock_isolate_workspaces;
+
+  switcher_isolate_workspaces = g_settings_get_boolean (self->shell_settings,
+    "current-workspace-only");
+  dock_isolate_workspaces = g_settings_get_boolean (self->dock_settings,
+    "isolate-workspaces");
+
+  if (switcher_isolate_workspaces != dock_isolate_workspaces)
+    {
+      g_settings_set_boolean (self->dock_settings, "isolate-workspaces",
+                              switcher_isolate_workspaces);
+    }
+}
+
+static void
 fixed_workspaces_changed_cb (CcMultitaskingPanel *self)
 {
   gboolean multi_workspaces, fixed_workspaces;
@@ -71,6 +92,7 @@ cc_multitasking_panel_finalize (GObject
   g_clear_object (&self->interface_settings);
   g_clear_object (&self->mutter_settings);
   g_clear_object (&self->shell_settings);
+  g_clear_object (&self->dock_settings);
   g_clear_object (&self->wm_settings);
 
   G_OBJECT_CLASS (cc_multitasking_panel_parent_class)->finalize (object);
@@ -101,12 +123,18 @@ cc_multitasking_panel_class_init (CcMult
   gtk_widget_class_bind_template_child (widget_class, CcMultitaskingPanel, workspaces_primary_display_radio);
   gtk_widget_class_bind_template_child (widget_class, CcMultitaskingPanel, workspaces_span_displays_radio);
 
+  gtk_widget_class_bind_template_child (widget_class, CcMultitaskingPanel, dock_monitors_isolation_radio);
+  gtk_widget_class_bind_template_child (widget_class, CcMultitaskingPanel, dock_each_monitor_radio);
+
   gtk_widget_class_bind_template_callback (widget_class, fixed_workspaces_changed_cb);
 }
 
 static void
 cc_multitasking_panel_init (CcMultitaskingPanel *self)
 {
+  GSettingsSchemaSource *schema_source = g_settings_schema_source_get_default ();
+  g_autoptr(GSettingsSchema) schema = NULL;
+
   g_resources_register (cc_multitasking_get_resource ());
 
   gtk_widget_init_template (GTK_WIDGET (self));
@@ -174,4 +202,32 @@ cc_multitasking_panel_init (CcMultitaski
       cc_illustrated_row_set_resource (self->active_screen_edges_row,
                                        "/org/gnome/control-center/multitasking/assets/active-screen-edges-rtl.svg");
     }
+
+  schema = g_settings_schema_source_lookup (schema_source,
+                                            "org.gnome.shell.extensions.dash-to-dock",
+                                            TRUE);
+  if (schema)
+    {
+      self->dock_settings = g_settings_new_full (schema, NULL, NULL);
+
+      g_signal_connect_object (self->shell_settings, "changed::current-workspace-only",
+                               G_CALLBACK (keep_dock_settings_in_sync), self,
+                               G_CONNECT_SWAPPED);
+      g_signal_connect_object (self->dock_settings, "changed::isolate-workspaces",
+                               G_CALLBACK (keep_dock_settings_in_sync), self,
+                               G_CONNECT_SWAPPED);
+
+      keep_dock_settings_in_sync (self);
+
+      if (g_settings_get_boolean (self->dock_settings, "isolate-monitors"))
+        gtk_check_button_set_active (self->dock_each_monitor_radio, TRUE);
+      else
+        gtk_check_button_set_active (self->dock_monitors_isolation_radio, TRUE);
+
+      g_settings_bind (self->dock_settings,
+                       "isolate-monitors",
+                       self->dock_each_monitor_radio,
+                       "active",
+                       G_SETTINGS_BIND_DEFAULT);
+    }
 }
--- a/panels/multitasking/cc-multitasking-panel.blp
+++ b/panels/multitasking/cc-multitasking-panel.blp
@@ -128,6 +128,27 @@ template $CcMultitaskingPanel: $CcPanel
             group: all_workspaces_radio;
           }
         }
+        Adw.PreferencesGroup monitor_isolation_group {
+          Adw.ActionRow {
+            activatable-widget: dock_monitors_isolation_radio;
+            title: _("Include apps from all _monitors");
+            use-underline: true;
+            [prefix]
+            CheckButton dock_monitors_isolation_radio {
+              valign: center;
+            }
+          }
+          Adw.ActionRow {
+            activatable-widget: dock_each_monitor_radio;
+            title: _("Include apps from _each monitor only");
+            use-underline: true;
+            [prefix]
+            CheckButton dock_each_monitor_radio {
+              valign: center;
+              group: dock_monitors_isolation_radio;
+            }
+          }
+        }
       }
     };
   };
